WIP: AArch64: AArch32: Add ARM 32-Bit Build and ARM/AArch64 Compiler to Travis CI Cross Compile Build

aarongraham9 opened the pull request:
Add ARM 32-Bit Build and ARM/AArch64 Compiler to Travis CI Cross Compile Build

Signed-off-by: Aaron Graham <aaron.graham@unb.ca>

aarongraham9 commented:
@JamesKingdon any thoughts why the ARM 32-Bit Travis build is failing to find the gcc executable based on these changes?

aarongraham9 commented:
I think I caught something. I pushed a potential fix. Let's see if it is still failing with this build (https://travis-ci.org/eclipse/omr/jobs/387811841). Thanks!

aarongraham9 commented:
Seems to be fixed. Thanks @JamesKingdon for your time!

aarongraham9 commented:
It seems that `AArch64` is, not-suprisingly, not implemented in the `jitbuilder` (when adding the `--enable-OMR_JITBUILDER` flag). I'm going to look at this this morning.

aarongraham9 commented:
Hey @JamesKingdon any thoughts as to why the ARM build (https://travis-ci.org/eclipse/omr/jobs/387954622) is failing when I added the `--enable-OMR_JITBUILDER` flag? It looks like there is a data type `TR::DataType` not defined?

JamesKingdon commented:
Sorry, I haven't looked at jitbuilder at all, so no clues here.

aarongraham9 commented:
Ok. I'll have to dig into it. Any recommendations @0xdaryl? I'd assume we wouldn't be happy with failing Travis builds in the meantime.

aarongraham9 commented:
Seems I'm still missing a Makefile somewhere as the build passes but none of the compiler is being built.

Leonardo2718 commented:
Haven't had a chance to look at this in close detail so this may not be the problem: in addition `--enable-OMR_JIT` do you also need `--enable-OMR_JITBUILDER` or `--enable-OMR_TEST_COMPILER`?

aarongraham9 commented:
Yes good catch! I was missing the `--enable-OMR_JITBUILDER` as I found out yesterday afternoon. Now I'm working on getting the platform to be detected and building the compiler code. Thanks!

aarongraham9 commented:
@0xdaryl here are the error occurrences in the logs you were asking about:
[Travis build failure log with: undeclared symbol `TR::DataType` on `ARM`](https://travis-ci.org/eclipse/omr/jobs/391746873#L1622)
and
[Travis build failure log: with re-definition of a macro `TR_ALLOC` on `AArch64`](https://travis-ci.org/eclipse/omr/jobs/391746874#L1679).

aarongraham9 commented:
Never-mind for the `AArch64` compile-time error regarding the `TR_ALLOC`. It's just a warning in the end and the build goes further. I guess I've been trained by most OMR and OpenJ9 code treating warnings as errors. Now I see that It's failing on finding the `aarch64/codegen/Instruction.hpp` header. Locally I've borrowed @knn-k's stub `aarch64/codegen/Instruction.hpp` header he wrote  for PR #2660 and I'm getting much further now.

The `ARM` error still stands however.

aarongraham9 commented:
This should be good for a review @0xdaryl, @JamesKingdon, @knn-k, and @jeanlego.

knn-k commented:
I have my version of compiler/aarch64/codegen/ARM64ConditionCode.hpp in #2704, which is not merged yet.  My version declares all the necessary symbols in its enum.  In addition, the type name is a little bit different: `TR_ARM64ConditionCode` vs `ARM64ConditionCode`.

@0xdaryl Should the type name contain `TR_` or not in this case?  The ARM32 version is declared as `ARMConditionCode`.

aarongraham9 commented:
@knn-k Ok; good catch! I think we should be consistent so I'll change ours to `TR_ARM64ConditionCode`.

aarongraham9 commented:
Unless we need to change `TR_ARM64ConditionCode` to `ARM64ConditionCode` this is good to merge @0xdaryl.

knn-k commented:
If the symbol should start with `TR_`, it is my #2704 that needs to be changed.

0xdaryl commented:
Yes, that is a good catch.  I would say `TR::ARM64ConditionCode`.  I missed that in 2704.

By rights this should maybe be `TR::AMD64::ConditionCode`, but I'm not yet prepared to go that far.  :-)

aarongraham9 commented:
Fixed @0xdaryl.

aarongraham9 commented:
Travis CI build finished. AppVeyor is still stalled as it seems to be lately. But as you say @0xdaryl we souldn't be affecting Windows builds anyway. This should be good to go.

0xdaryl commented on compiler/aarch64/codegen/OMRTreeEvaluator.hpp:
```diff
@@ -0,0 +1,789 @@
+namespace TR { class Instruction; }
+namespace TR { class Node; }
+namespace TR { class Register; }
+namespace TR { class RegisterDependencyConditions; }
```
Can you trim this down to the minimum number of includes and forward declarations you need for this file?

0xdaryl commented on compiler/aarch64/codegen/TreeEvaluatorTable.hpp:
```diff
@@ -0,0 +1,748 @@
+/*
+ * This table is #included in a static table.
+ * Only Function Pointers are allowed.
+ */
```
Can you create a `TR::TreeEvaluator::unImpOpEvaluator` like the other platforms do that essentially asserts?  Each IL opcode in this table should then map to that `TR::TreeEvaluator::unImpOpEvaluator` until it is actually implemented.  It is far easier to determine what evaluators are complete and what aren't by looking at a table annotated that way.

0xdaryl commented on compiler/aarch64/codegen/TreeEvaluatorTable.hpp:
```diff
@@ -0,0 +1,748 @@
+ * This table is #included in a static table.
+ * Only Function Pointers are allowed.
+ */
+    TR::TreeEvaluator::BadILOpEvaluator ,		// illegal op hopefully help with uninitialized nodes
```
Rather than using comments describing the operation, can you actually list the IL OpCode (perhaps in addition to the comment) like the other architectures do.  It makes it easier to index into this table by searching for the opcode you're interested in.  At some point, there may not be a 1-1 mapping between IL evaluator names and IL opcodes (e.g., you may create an "integerAddEvaluator" that handles multiple sizes of inputs that is shared across multiple IL opcodes) and in that case its easier to search the table for the IL opcode/evaluator pairing by using the IL opcode name.

jeanlego replied:
this was a cleaner hack to get it to build. Otherwise it fails static assertion (nb of IL op == nb of Ops in tree evaluator table)
I could just ifndef (TR_TARGET_AARCH64) the static assertion rather and loose the tree evaluator stuff and reassess later so this gets merged in and less bloat on the commit? thoughts ? @0xdaryl

0xdaryl commented on compiler/aarch64/codegen/ARM64ConditionCode.hpp:
```diff
@@ -0,0 +1,30 @@
+typedef enum 
+{
+} 
+ARM64ConditionCode;
```
Why does this need to be part of this commit if its not used and @knn-k has another implementation coming in #2744?

jeanlego replied:
without this it does not compile

0xdaryl commented on compiler/ras/Debug.hpp:
```diff
@@ -594,6 +594,9 @@ class TR_Debug
 #endif
+#if defined(TR_TARGET_ARM64)
+   void print(TR::FILE *pOutFile, TR::UnresolvedDataSnippet * snippet);
+#endif
```
Why is this needed, and where is it implemented?

jeanlego replied:
without this it does not compile. We could have the proper file stub or just have a temporary definition there. chose the later

jeanlego commented on compiler/env/defines.h:
```diff
@@ -67,6 +67,9 @@
 #  define ARCH_ARM     104
 #endif
+#ifndef ARCH_ARM64
+#  define ARCH_ARM64     104
```
this was a typo! oups

jeanlego commented on jitbuilder/build/toolcfg/gnu/common.mk:
```diff
@@ -351,12 +351,20 @@ ifeq ($(HOST_BITS),32)
+        ifneq ($(HOST_ARCH),aarch64)
+            SOLINK_FLAGS+=-m64
+        endif
+    endif
```
-m64 or -m32 is not a valid flag for cross compiling

jeanlego commented on compiler/env/OMRObjectModel.hpp:
```diff
@@ -39,6 +39,9 @@ class TR_OpaqueClassBlock;
 namespace TR { class Node; }
+#ifdef TR_TARGET_ARM64
+namespace TR { class DataType; }
+#endif
```
this was needed because we don't have all the stub necessary yet for aarch64. This allows it to compile without having to stub out as much

aarongraham9 commented:
Can you re-run the Travis build @0xdaryl? It built successfully on the CAS-Atlantic Travis https://travis-ci.org/CAS-Atlantic/omr/builds/404551368 but the cross-compile toolchain just timed out for the ARM (32-Bit) part of the matrix build in the OMR Travis https://travis-ci.org/eclipse/omr/builds/404551383.

jeanlego commented:
Looks good to me :)

aarongraham9 commented:
Addressed review comments. Split commits and updated commit messages. Rebased off of the `eclipse/omr:master` branch@ eclipse/omr@5d8e52c1de6d9dbde10b85249f15d1517d9db773 to include the Travis fix from @Leonardo2718 and merged PR(s) from @knn-k.

Travis CI Cross-Compile Build passed successfully.

Unless there are any further review comments this should be good to go @0xdaryl.

aarongraham9 commented:
Addressed review comments. Split commits and updated commit messages. Rebased off of the `eclipse/omr:master` branch(eclipse/omr@5d8e52c1de6d9dbde10b85249f15d1517d9db773) to include the Travis fix from @Leonardo2718 and merged PR(s) from @knn-k.

Travis CI Cross-Compile and AppVeyor builds passed successfully.

Unless there are any further review comments this should be good to go @0xdaryl.
